crates/ic-ingress-validator-util/src/delegation_validator.rs

```diff
@@ -0,0 +1,382 @@
+            InvalidBasicSignature(err) => write!(f, "Invalid basic signature: {}", err),
+            InvalidCanisterSignature(err) => write!(f, "Invalid canister signature: {}", err),
+            InvalidPublicKey(err) => write!(f, "Invalid public key: {}", err),
+            WebAuthnError(msg) => write!(f, "{}", msg),
+            DelegationTargetError(msg) => write!(f, "{}", msg),
+            DelegationTooLongError { length, maximum } => write!(
+                f,
+                "Chain of delegations is too long: got {} delegations, but at most {} are allowed",
+                length, maximum
+            ),
+            DelegationContainsCyclesError { public_key } => write!(
+                f,
+                "Chain of delegations contains at least one cycle: first repeating public key encountered {}",
+                hex::encode(public_key)
+            ),
+        }
+    }
+}
+
+// Check if ingress_expiry is within a proper range with respect to the given
+// time, i.e., it is not expired yet and is not too far in the future.
+fn validate_ingress_expiry(
+    request_expiry: u64,
+    current_time: Time,
+) -> Result<(), RequestValidationError> {
+    let provided_expiry = Time::from_nanos_since_unix_epoch(request_expiry);
+    let min_allowed_expiry = current_time;
+    // We need to account for time drift and be more forgiving at rejecting ingress
+    // messages due to their expiry being too far in the future.
+    let max_expiry_diff = MAX_INGRESS_TTL
+        .checked_add(PERMITTED_DRIFT_AT_VALIDATOR)
+        .ok_or_else(|| {
+            InvalidIngressExpiry(format!(
+                "Addition of MAX_INGRESS_TTL {MAX_INGRESS_TTL:?} with \
+                PERMITTED_DRIFT_AT_VALIDATOR {PERMITTED_DRIFT_AT_VALIDATOR:?} overflows",
+            ))
+        })?;
+    let max_allowed_expiry = min_allowed_expiry
+        .checked_add(max_expiry_diff)
+        .ok_or_else(|| {
+            InvalidIngressExpiry(format!(
+                "Addition of min_allowed_expiry {min_allowed_expiry:?} \
+                with max_expiry_diff {max_expiry_diff:?} overflows",
+            ))
+        })?;
+    if !(min_allowed_expiry <= provided_expiry && provided_expiry <= max_allowed_expiry) {
+        let msg = format!(
+            "Specified ingress_expiry not within expected range: \
+             Minimum allowed expiry: {}, \
+             Maximum allowed expiry: {}, \
+             Provided expiry:        {}",
+            min_allowed_expiry, max_allowed_expiry, provided_expiry
+        );
+        return Err(InvalidIngressExpiry(msg));
+    }
+    Ok(())
+}
+
+// Verifies that the user id matches the public key.  Returns an error if not.
+fn validate_user_id(sender_pubkey: &[u8], id: PrincipalId) -> Result<(), RequestValidationError> {
+    if id == PrincipalId::new_self_authenticating(sender_pubkey) {
+        Ok(())
+    } else {
+        Err(UserIdDoesNotMatchPublicKey(
+            UserId::new(id),
+            sender_pubkey.to_vec(),
```

Comment:
not sure if the error display impl allows for the hex encoding of these bytes but unless you want to be able to match on this error and make use of the raw bytes of the pubkey in the error handling code, why not just do the display-encoding here and then interpolate that value into the error message without encoding it there?